File: Data Structures & Algorithms/implement-prefix-tree/submission-3.go
type PrefixTree struct {
    children [26]*PrefixTree
    isEnd    bool
}

func Constructor() PrefixTree {
    return PrefixTree{}
}

func (t *PrefixTree) Insert(word string) {
    node := t
    for _, c := range word {
        i := c - 'a'
        if node.children[i] == nil {
            node.children[i] = &PrefixTree{}
        }
        node = node.children[i]
    }
    node.isEnd = true
}

func (t *PrefixTree) find(word string) *PrefixTree {
    node := t
    for _, c := range word {
        i := c - 'a'
        if node.children[i] == nil {
            return nil
        }
        node = node.children[i]
    }
    return node
}

func (t *PrefixTree) Search(word string) bool {
    node := t.find(word)
    return node != nil && node.isEnd
}

func (t *PrefixTree) StartsWith(prefix string) bool {
    return t.find(prefix) != nil
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-3.java
class TrieNode {
    TrieNode[] children = new TrieNode[26];
    boolean isEnd = false;
}

class WordDictionary {
    private TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }

    public void addWord(String word) {
        TrieNode cur = root;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (cur.children[idx] == null) {
                cur.children[idx] = new TrieNode();
            }
            cur = cur.children[idx];
        }
        cur.isEnd = true;
    }

    public boolean search(String word) {
        return dfs(root, word, 0);
    }

    private boolean dfs(TrieNode node, String word, int i) {
        if (node == null) return false;
        if (i == word.length()) {
            return node.isEnd;
        }
        char c = word.charAt(i);
        if (c == '.') {
            for (TrieNode child : node.children) {
                if (child != null && dfs(child, word, i + 1)) {
                    return true;
                }
            }
            return false;
        } else {
            return dfs(node.children[c - 'a'], word, i + 1);
        }
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */